### ISSUE TITLE
Multi-touch?

### ISSUE THREAD

**Issue by greggman** [ocornut/imgui#165]
I don't know it this makes sense but lets say I wanted 5 vertical sliders and on a touch based system I wanted to be able to put 5 fingers on the screen and move each slider at the same time.

Is that something that's already possible? If not ever considered how the code would have to be refactored to support it?

**Comment by ocornut** [comment]
ARGH.

Not totally crazy. Basically all those internal fields:

'''
ImGuiID                 ActiveId;
ImGuiID                 ActiveIdPreviousFrame;
bool                    ActiveIdIsAlive;
bool                    ActiveIdIsFocusedOnly;
ImGuiWindow*            MovedWindow;          
[...]
ImGuiTextEditState      InputTextState;
ImGuiID                 SliderAsInputTextId;
[...]
ImGuiID                 ActiveComboID;
float                   ScrollbarClickDeltaToGrabCenter;
'''

Would have to be moved into a structure, put into an array and handled as such (which a bunch of helpers, tracking touch id). My intuition is that it isn't worth it at the moment because:

- I can envision lots of bugs and maintenance with multi-touch edge-case behaviours.
- It would make every bit of the code slightly bulkier and slower.
- Last time I tested it on the Vita for Tearaway was about 18 months ago it did work with hover=touch, but there may be some bugs to fix.
- Imprecise inputs are really badly supported by ImGui. You can increase style.TouchPadding, but here's no sorting and everything is one-pass so first widget to run (typically rendering left-to-right top-to-bottom) can steal the touch. So typically if you increase style.TouchPadding you want to increase style.ItemSpacing accordingly. Either way it's never going to be ideal for touch inputs.

My worry at the moment is that we have so many other things to work on to consider this. Maybe next year when we lots of major features done and polished it would make sense, but at the moment my intuition is that it'll get in the way. All the internal widget code will probably have been cleaned up by then it may also be easier.

Note that it'd be possible to implement a subset of features e.g. ImGui::MultititouchSlider() externally, using some of the function that ImGui expose and using the ImDrawList API, managing it yourself, and it may be enough for your app.

**Comment by greggman** [comment]
Given that nearly every Windows machine now comes with a touch display it seems ignoring touch might start to seem like a bad decision at some point? Or not. 

I don't know what the various native APIs are like. The MS PointerEvents JavaScript API tracks the fingers for you. Following that model if a button sees it's been pressed by finger 3 then at one level it only cares about movement for finger 3 and release by finger 3. 

Maybe working backward, if the finger knows it was in a certain button last time I can use that to make most of this stuff transparent? Just thinking out loud. I haven't looked at the code.

**Comment by ocornut** [comment]
Yes I am making it a bit more complicated that it is. The fields above are more or less the state that widgets refers to. It would work like in your middle paragraph. Put those in an array, use the widget ID to check if it's active and find the corresponding structure to use. Also need to rework where it is getting inputs from.

The thing is that Imgui is useful for tools/debug and I don't think this is the type of situation where multi-touch is mostly commonly useful. Let me rephrase that: I think we should do it eventually, but it'd be rather low in the priority list at the moment. It feels like one of those things that's not actually highly desired by users.

It's good to keep that in mind though and may come sooner depending on what I am working on.

**Comment by ocornut** [comment]
I'm closing this now. Not that it isn't a legit enhancement but I also know you asked out of curiosity (and not actually being a ImGui user) and I've got so much on my plate that I need to keep the issue list manageable to real world priorities.
